### PR TITLE
Fix geth tests

### PR DESCRIPTION
**Pull request by cabrador**
## Description

This PR fixes file iteration by
1) correctly adding information to each tests such as description
2) adding expected error - some tests are designed to fail
and `blockEnviroment` of gethTests by correctly calculating each value depending on which fork is test running.

#1159 should be merged before this PR

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: ethtest/state_test_context.go [hunk 1]
```diff
@@ -23,28 +23,32 @@ import (
...
-		fork:       fork,
-		path:       stJson.path,
-		number:     number,
-		env:        env,
-		inputState: stJson.Pre,
-		msg:        msg,
-		rootHash:   rootHash,
+		fork:          fork,
+		path:          stJson.path,
+		postNumber:    postNumber,
```
Comment (wsodsong) [inline ethtest/state_test_context.go:33]: nit: it would be helpful if you can add short descriptions of these variables. What is postNumber?

**Review by wsodsong — DISMISSED** [review]

File: ethtest/block_enviroment.go [hunk 2]
```diff
@@ -20,33 +20,56 @@ import (
...
 func (s *stBlockEnvironment) GetDifficulty() *big.Int {
-	return s.Difficulty.Convert()
+	var difficulty *big.Int
+	if s.Difficulty != nil {
+		difficulty = s.Difficulty.Convert()
+	}
+
+	if s.chainCfg.IsLondon(new(big.Int)) && s.Random != nil {
+		difficulty = big.NewInt(0)
+	}
```
Comment (wsodsong) [inline ethtest/block_enviroment.go:61]: This code block read as if a block is London or later and `Random` is given, set difficult to 0.
Just want to confirm if `London` fork correct, not `Paris` (the merge)?

File: ethtest/test_case_splitter_test.go [hunk 3]
```diff
@@ -9,15 +9,24 @@ import (
...
 	stJson := CreateTestStJson(t)
-	d := TestCaseSplitter{jsons: []*stJSON{stJson}, log: logger.NewLogger("info", "test-case-splitter-test")}
-	for _, testCase := range d.SplitStateTests() {
-		msg := testCase.GetMessage()
+	d := TestCaseSplitter{
+		jsons:        []*stJSON{stJson},
+		log:          logger.NewLogger("info", "test-case-splitter-test"),
+		chainConfigs: make(map[string]*params.ChainConfig),
+	}
+	tests, err := d.SplitStateTests()
```
Comment (wsodsong) [inline ethtest/test_case_splitter_test.go]:
```suggestion
	testsuite := TestCaseSplitter{
		jsons:        []*stJSON{stJson},
		log:          logger.NewLogger("info", "test-case-splitter-test"),
		chainConfigs: make(map[string]*params.ChainConfig),
	}
	tests, err := testsuite.SplitStateTests()
```

File: ethtest/state_test_context.go [hunk 4]
```diff
@@ -23,28 +23,34 @@ import (
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core"
 	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/params"
 )
 
-func newStateTestTxContest(stJson *stJSON, env *stBlockEnvironment, msg *core.Message, rootHash common.Hash, fork string, number int) txcontext.TxContext {
+func newStateTestTxContest(stJson *stJSON, msg *core.Message, post stPost, chainCfg *params.ChainConfig, fork string, postNumber int) txcontext.TxContext {
```
Comment (wsodsong) [inline ethtest/state_test_context.go]:
```suggestion
func newStateTestTxContext(stJson *stJSON, msg *core.Message, post stPost, chainCfg *params.ChainConfig, fork string, postNumber int) txcontext.TxContext {
```

File: ethtest/test_case_splitter.go [hunk 5]
```diff
@@ -59,48 +69,74 @@ func sortForks(log logger.Logger, cfgForks []string) (forks []string) {
...
 
-				dividedTests = append(dividedTests, newStateTestTxContest(stJson, env, msg, post.RootHash, fork, number))
+				if fork == "Paris" {
+					fork = "Merge"
+				}
+				ctx := newStateTestTxContest(stJson, msg, post, chainCfg, fork, postNumber)
+				dividedTests = append(dividedTests, Transaction{
+					fork,
+					ctx,
+				})
```
Comment (wsodsong) [inline ethtest/test_case_splitter.go]: I feel `ctx` is too generic. 
```suggestion
				txCtx := newStateTestTxContest(stJson, msg, post, chainCfg, fork, postNumber)
				dividedTests = append(dividedTests, Transaction{
					fork,
					txCtx,
				})
```

File: ethtest/test_case_splitter.go [hunk 6]
```diff
@@ -59,48 +69,74 @@ func sortForks(log logger.Logger, cfgForks []string) (forks []string) {
...
 
-	return dividedTests
+	return dividedTests, err
+}
+
+func (s *TestCaseSplitter) getChainConfig(fork string) (*params.ChainConfig, error) {
+	if cfg, ok := s.chainConfigs[fork]; ok {
+		return cfg, nil
+	}
+	cfg, _, err := tests.GetChainConfig(fork)
```
Comment (wsodsong) [inline ethtest/test_case_splitter.go:135]: Is there a different between Ethereum chain config form our version of go-ethereum to the one used in `tests` package?

**Review by wsodsong — CHANGES_REQUESTED** [review]

File: ethtest/block_enviroment.go (hunk 2 above)
Comment (cabrador) [reply ethtest/block_enviroment.go:61]: Yes, `London` is correct. It comes from [this](https://github.com/ethereum/go-ethereum/blob/ec596e06a58f3c4c4bac0262ebfc57539c3e5fd1/tests/state_test_util.go#L288) line.

File: ethtest/test_case_splitter.go (hunk 6 above)
Comment (cabrador) [reply ethtest/test_case_splitter.go:135]: This is a special `chainConfig` created for the `tests` package. It aligns with the block setup in the `stBlockEnvironment`. You can check out the func [here](https://github.com/ethereum/go-ethereum/blob/ec596e06a58f3c4c4bac0262ebfc57539c3e5fd1/tests/state_test_util.go#L142).

The key part is this [map](https://github.com/ethereum/go-ethereum/blob/ec596e06a58f3c4c4bac0262ebfc57539c3e5fd1/tests/init.go#L30) used on line 148.

Every time `IsSomeFork()` (ie `IsLondon()`, `IsCancun()`, etc...) is used, it specifically asks for values from this map.

File: ethtest/test_case_splitter_test.go (hunk 3 above)
Comment (cabrador) [reply ethtest/test_case_splitter_test.go]: mb - I forgot to rename this. I will use `splitter` if you are fine with this name

**Review by wsodsong — APPROVED** [review]
Looks good! Thank you for the fix.

**Pull request merged**